LineChart: Allow to hide graph series

Gatos90 opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->


## Description
<!-- Describe your changes in detail any why -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310 -->

## How Has This Been Tested?
<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

Commit:
adds Visibility to Chatseries, for hide, show line

Commit:
LineChart: Hide Line depending on Visibility

Gatos90 committed:
Update Line.razor.cs

jacob7395 commented:
Would you be able to update the legend to indicate what is currently displayer, maybe gray out or put a line though not displayed values?

jacob7395 commented on src/MudBlazor/Components/Chart/Charts/Line.razor.cs:
```diff
@@ -201,14 +201,15 @@ private void GenerateChartLines(int lowestHorizontalLine, double gridYUnits, dou
                 }
-
+                if (_series[i].IsVisible)
+                {
```
This should probably be indented like the other sections in this file

Commit:
LineChart: Add Line through Legend of hidden Line

Gatos90 commented:
Now the Legend has an Indicator for the hidden parts:
![image](https://github.com/MudBlazor/MudBlazor/assets/54571782/2480b513-86c6-4f80-afd0-75a5049f3de6)

Commit:
added a Test for the isVisible

ScarletKuro commented:
lgtm, @henon any comments?

henon commented:
I think the hidden graphs schould be entirely hidden in the legend as well. This would be most consequent. Hidden data is hidden both in graph and legend. But then wouldn't that bring the entire PR ad absurdum, wouldn't it? Because you could just remove the entire `ChartSeries` and re-render the chart. 

Like so:
https://try.mudblazor.com/snippet/mamnbQlYBxtKTnRF

henon commented:
Ah, sorry I see what you want to do, you want to switch between graphs in the legend.

henon commented on src/MudBlazor/Components/Chart/Parts/Legend.razor:
```diff
@@ -8,7 +8,11 @@
         {
             <div class="mud-chart-legend-item" @onclick=@(()=>{ if (MudChartParent!=null) { MudChartParent.SelectedIndex=item.Index; }}) @onclick:stopPropagation=@(MudChartParent!=null)>
                 <span class="mud-chart-legend-marker" style="@($"background-color:{MudChartParent.ChartOptions.ChartPalette.GetValue(item.Index % ChartOptions.ChartPalette.Count())}")"></span>
-                <MudText Typo="Typo.body2" Inline="true">@item.Labels</MudText>
+                @if (item.IsVisible) {
+                    <MudText Typo="Typo.body2" Inline="true">@item.Labels</MudText>}
+                else {
+                    <MudText Typo="Typo.body2" Inline="true" Style="text-decoration: line-through;">@item.Labels</MudText>
+                }
             </div>
         }
     </div>
```
Rather than this we need a RenderFragment parameter for the legend. You could for instance add check buttons or radio buttons to the legend or whatever. In any case, the current solution is too specific and will surely not satisfy everyone.

henon commented:
A simple and easy default behavior would also be to set the z-index of the graph that was last clicked in the legend to a value higher than all the other graphs. 

But of course legend render fragment and visibility are also good features.

Commit:
Charts: Added checkboxes and option for activate hiding lines

Gatos90 commented:
Introduced a variable toggle that enables or disables the ability to show or hide chart lines. When this feature is activated, checkboxes appear next to each line. Users can interact with these checkboxes to selectively display or conceal specific lines on the chart. I also added an example how to use it in the Docs.
![image](https://github.com/MudBlazor/MudBlazor/assets/54571782/48e905bc-41aa-4d5d-a10b-6eb71a27bcf3)
![image](https://github.com/MudBlazor/MudBlazor/assets/54571782/331f84b8-8426-44fc-90bd-b22e9e8518f8)
![image](https://github.com/MudBlazor/MudBlazor/assets/54571782/8968bffb-20e9-4fb5-a677-9ae4e2e6c3d3)

henon commented on src/MudBlazor/Components/Chart/Charts/Line.razor.cs:
```diff
@@ -201,21 +201,35 @@ private void GenerateChartLines(int lowestHorizontalLine, double gridYUnits, dou
+            if (series != null)
+            {
+                series.IsVisible = legend.IsVisible;
+                Console.WriteLine(series.IsVisible);
```
Please remove the `Console.WriteLine`

henon commented:
What would you say to the idea of replacing the circle with a colored checkbox if the `CanHideSeries` parameter is true? Wouldn't that look even better?

ScarletKuro commented on src/MudBlazor/Components/Chart/Svg/SvgLegend.cs:
```diff
@@ -1,10 +1,21 @@
+        public bool IsVisible {get; set;} = true;
+        public EventCallback<SvgLegend> OnVisibilityChanged { get; set; }
+
+        public async Task HandleCheckboxChange(bool isVisible)
```
Please rename to `HandleCheckboxChangeAsync`.

ScarletKuro commented on src/MudBlazor/Components/Chart/Svg/SvgLegend.cs:
```diff
@@ -1,10 +1,21 @@
         public int Index { get; set; }
         public string Labels { get; set; }
         public string Data { get; set; }
+        public bool IsVisible {get; set;} = true;
```
Wrong formatting, there should be spaces `{ get; set; }`

Commit:
Changed Names, and Checkboxes to work with LineChart

Gatos90 commented:
![image](https://github.com/MudBlazor/MudBlazor/assets/54571782/44114a08-af94-4b5f-9247-6bdd0eb3e210)
![image](https://github.com/MudBlazor/MudBlazor/assets/54571782/95221c29-5349-4d58-9559-33b7c0424887)

Added to make this work, custom colors to Checkbox:
![image](https://github.com/MudBlazor/MudBlazor/assets/54571782/53a32c58-7be3-4a37-b1e5-e46f7ee6842f)
![image](https://github.com/MudBlazor/MudBlazor/assets/54571782/e31cb3a7-1810-4f0d-bbc9-48b2129ba2d5)

Commit:
changed test

ScarletKuro commented on src/MudBlazor/Components/Chart/Parts/Legend.razor:
```diff
@@ -4,11 +4,21 @@
 {
     <div @attributes="UserAttributes" class="mud-chart-legend">
-        @foreach (var item in Data)
+        @foreach (var item in Data)        
```
Extra spaces after that for each. Otherwise LGTM

Gatos90 replied:
removed

Gatos90 committed:
Removed extra spaces

henon commented:
I am sorry that I have to raise this objection, but is there no other way to override the color of the check boxes? We want to support custom colors in MudBlazor components but we'll probably implement it very differently.

Gatos90 commented:
I've thought through a number of solutions, and I think the current method—using the Color enum together with an extra check for custom colors—is the most efficient. It makes sure that the default colors are the standard colors, but it also gives you the option to utilize custom colors if necessary. I think this approach is a good compromise between functionality and minimalism because it makes modifications easy to make and less invasive. Furthermore, this will be simple to alter if mudblazors components have custom colors. Greating

Commit:
Colors now changable, without Custom Colors for Checkboxes.

Commit:
Merge branch 'feature/lineChartVisibility' of https://github.com/Gatos90/MudBlazor into feature/lineChartVisibility

Commit:
removed Custom Color and added it to Legend an a class

henon commented:
Oh nice, your solution is much better than what I proposed!

Gatos90 commented:
Without your idea, i would have not found this solution. So thank U! And thx for merging!
